src/parallel.jl

```diff
@@ -0,0 +1,143 @@
+        return false
+    end
+end
+
+function parallel_spdmp_inner!(events, partition, ti, inner, G, G1, G2, ∇ϕ, t, x, θ, Q, c, b, t_old,
+    F::Union{ZigZag,FactBoomerang}, (factor, adapt), args...)
+   n = length(x)
+   acc = num = 0
+   while true
+        num += 1
+        ii, t′ = peek(Q[ti])
+        i = partition(ti, ii)
+        if !inner[i] # need neighbours at t′
+           return false, event(i, t, x, θ, F), i, t′, acc, num
+        end
+
+        success = parallel_innermost!(partition, G, G1, G2, ∇ϕ, i, t, x, θ, t′, Q, c, b, t_old, F, (factor, adapt), args...)
+
+        success || continue
+        acc += 1
+        #return true, event(i, t, x, θ, F), i, t′, acc, num
+        push!(events[ti], event(i, t, x, θ, F))
+   end
+end
+
+function parallel_spdmp(partition, ∇ϕ, t0, x0, θ0, T, c, G, F::Union{ZigZag,FactBoomerang}, args...;
+    factor=1.8, adapt=false)
+    nthr = length(partition)
+    n = length(x0)
+    t′ = t0
+    t = fill(t′, size(θ0)...)
+    t_old = copy(t)
+    if G === nothing
+        G = [i => rowvals(F.Γ)[nzrange(F.Γ, i)] for i in eachindex(θ0)]
+    end
+    inner = [all(j -> partition(j)[1] == partition(i)[1], js) for  (i,js) in G]
+ #   @show inner
+    G1 = [i => rowvals(F.Γ)[nzrange(F.Γ, i)] for i in eachindex(θ0)]
+
+    @assert all(a.second ⊇ b.second for (a,b) in zip(G, G1))
+
+    G2 = [i => setdiff(union((G1[j].second for j in G1[i].second)...), G[i].second) for i in eachindex(G1)]
+    x, θ = copy(x0), copy(θ0)
+
+    Q = [SPriorityQueue{Int,Float64}() for thr in 1:nthr]
+    b = [ab(G1, i, x, θ, c, F) for i in eachindex(θ)]
+    for i in eachindex(θ)
+        q1, q2 = partition(i)
+        enqueue!(Q[q1], q2 => poisson_time(b[i], rand()))
+    end
+    Ξ = Trace(t0, x0, θ0, F)
+    task = Vector{Task}(undef, length(partition))
+    evtime = zeros(nthr)
+    perm = collect(1:nthr)
+    events = [[event(1, 0., x, θ, F)] for ts in each(partition)]
+    res = [(true, event(1, 0., x, θ, F), 1, 1.0, 1, 1) for ts in each(partition)]
+    parallel_spdmp_loop(t′, T, task, evtime, perm, res, Ξ, events, partition, inner, G, G1, G2, ∇ϕ, t, x, θ, Q,
+    c, b, t_old, F, (factor, adapt), args...)
+end
+function parallel_spdmp_loop(t′, T, task, evtime, perm, res, Ξ, events, partition, inner, G, G1, G2, ∇ϕ, t, x, θ, Q,
+    c, b, t_old, F, (factor, adapt), args...)
+    num = acc = 0
+    run = runs = 0
+    while t′ < T
+        @threads for ti in each(partition) # can be parallel
+            resize!(events[ti], 0)
+            res[ti] = parallel_spdmp_inner!(events, partition, ti, inner, G, G1, G2, ∇ϕ, t, x, θ, Q,
+                    c, b, t_old, F, (factor, adapt), args...) 
+        end
+        
+        for ti in each(partition)
+            run += res[ti][end]
+            runs += 1
+            #println(res[ti][end])
+            append!(Ξ.events, events[ti])
+            evtime[ti] = res[ti][4]
+        end
+        sortperm!(perm, evtime, alg=InsertionSort)
+        for ti in perm
```

Comment:
Imagine the following situation: I am at `T = 2` and need to check the neighbour. Neighbour is at `T=1`, looks at my boundary. I can not immediately use that, because the neighbour might _stop again_ and change the boundary before `T=2`, say at `T=1.5`

Reply:
Issue: Neighbours might have repeatedly problematic cases in your past